src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+        Versioned,
+    },
+    contracts_common::{AccountAddress, SignatureThreshold},
+    curve_arithmetic::Curve,
+    id::types::{
+        AccountCredentialWithoutProofs, AccountKeys, Attribute, InitialAccountData,
+        PublicCredentialData, VerifyKey,
+    },
+};
+use ed25519_dalek::{SigningKey, VerifyingKey};
+use sha2::Digest;
+use std::collections::BTreeMap;
+
+/// Errors that can occur during generating or verifying account signatures.
+#[derive(thiserror::Error, Debug)]
+pub enum SignatureError {
+    #[error("Network error: {0}")]
+    QueryError(#[from] QueryError),
+    #[error(
+        "Indices do not exist on chain for credential index `{credential_index}` and key index \
+         `{key_index}`"
+    )]
+    MissingIndicesOnChain {
+        credential_index: u8,
+        key_index:        u8,
+    },
+    #[error("The indices in the maps do not match")]
+    MismatchMapIndices,
+    #[error(
+        "The public key and the private key in the `account_keys` map do not match for credential \
+         index `{credential_index}` and key index `{key_index}`"
+    )]
+    MismatchPublicPrivateKeys {
+        credential_index: u8,
+        key_index:        u8,
+    },
+    #[error(
+        "The public key on chain `{expected_public_key:?}` and the public key \
+         `{actual_public_key:?}` in the signature map do not match for credential index \
+         `{credential_index}` and key index `{key_index}`"
+    )]
+    MismatchPublicKeyOnChain {
+        credential_index:    u8,
+        key_index:           u8,
+        expected_public_key: Box<VerifyingKey>,
+        actual_public_key:   Box<VerifyingKey>,
+    },
+}
+
+/// Message types that can be signed and verified. The Concordium wallets can
+/// either sign binary or text messages.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Message<'a> {
+    BinaryMessage(&'a [u8]),
+    TextMessage(&'a str),
+}
+
+/// Account signatures are constructed similarly to transaction signatures. The
+/// only difference is that account signatures use as account nonce the value 0
+/// when signing. This is to ensure that the user does not accidentally sign a
+/// valid transaction. As such account signatures can be used to sign messages
+/// in wallets that can then be verified in a smart contract or by a backend.
+///
+/// Account signatures should be thought of as a nested map, indexed on the
+/// outer layer by credential indices, and the inner map maps key indices to
+/// [`Signature`]s.
+#[derive(Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[repr(transparent)]
+pub struct AccountSignatures {
+    pub sigs: BTreeMap<u8, CredentialSignatures>,
+}
+
+#[derive(Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[repr(transparent)]
+pub struct CredentialSignatures {
+    pub sigs: BTreeMap<u8, Signature>,
+}
+
+impl AccountSignatures {
+    /// Create the account signature type from a single signature at credential
+    /// and key indices 0.
```

Comment:
An account with only a credential/key index 0 is a single-signer account (the most common type of account). 
These accounts have exactly one private-public key pair associated to them in contrast to multi-sig accounts. 

The wording `singleton` is used because we have similar function constructors for single-signer accounts in our base libraries.
https://github.com/Concordium/concordium-base/blob/13048aa52e424f854790945265054412e1d24b3d/rust-src/concordium_base/src/id/types.rs#L2203

Reply:
Do we have some documentation on this anywhere? Might be good to link to it or at least mention it here if not.

Reply:
There are a couple of comments mentioning that these functions exist to simplify the interactions for `single-signer` accounts in the PR already.

We have a good tutorial about signing and signature maps as well:
https://developer.concordium.software/en/mainnet/smart-contracts/tutorials/sponsoredTransactions/index.html#sponsoredtransactions